keel-core/src/test/kotlin/com/netflix/spinnaker/keel/services/ApplicationServiceTests.kt

```diff
@@ -123,250 +116,401 @@ class ApplicationServiceTests : JUnit5Minutests {
     }
 
     before {
-      clock.reset()
-    }
-
-    after {
-      deliveryConfigRepository.dropAll()
-      artifactRepository.dropAll()
-      resourceRepository.dropAll()
-      clearAllMocks()
+      every { repository.getDeliveryConfigForApplication(application) } returns deliveryConfig
     }
 
```

Comment:
The only thing I'm slightly concerned with in this PR is the changes to this class. It took a lot of work to replicate real situations and test both the db and the transforming for this presentation layer. It's really hard to reproduce these situations in practice (locally to test, or in a test to reproduce behavior), and I think switching to mocking everything here does us a disservice. 

Do you think there's a way that we could switch these tests to use the sql repositories, or perhaps extract them to a new test focused specifically on testing both the extraction from the database and the rendering for the api layer? That doesn't feel super "nice" to me, but I'd really rather have this way to test complicated situations

Reply:
It's certainly a possibility. One thing I'd like to retain would be the hierarchical context structure I set up here, which I think makes the test a lot easier to follow.

Reply:
I think it's scary because the tests here now look so different, but I really like the way that @robfletcher structured the new tests and think they're pretty readable once you stop comparing them to the old implementation and look at them on their own.